### PR TITLE
aggregate_list in bounded context

### PR DESCRIPTION
**Pull request by hallgren**
Stared to implement the DDD concept Bounded Context. The aggregate_list will return all AggregateRoots in a BoundedContext module.

The idea is to use it together with the global configuration map_types to make it easier to list all aggregates in a bounded context and add then to an event store.

### CODE REVIEW

File: lib/sandthorn/bounded_context.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    extend ClassMethods
+    def self.included( other )
+      other.extend( ClassMethods )
+    end
+    
+
+  end
+end
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Missing newline

File: lib/sandthorn/bounded_context.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  module BoundedContext
+
+    module ClassMethods
+      def aggregate_list()
+        @aggregate_list = p_aggregate_list(self)
+      end
+
+      private
+      def p_aggregate_list(bounded_context_module)
+        return [] unless bounded_context_module.respond_to?(:constants)
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Shouldn't we just let this fail?

File: lib/sandthorn/bounded_context.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+module Sandthorn
+  module BoundedContext
+
+    module ClassMethods
+      def aggregate_list()
+        @aggregate_list = p_aggregate_list(self)
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Maybe we want to memoize this?

File: lib/sandthorn/bounded_context.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+module Sandthorn
+  module BoundedContext
+
+    module ClassMethods
+      def aggregate_list()
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Remove parentheses.

File: lib/sandthorn/bounded_context.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        classes = bounded_context_module.constants.map(&bounded_context_module.method(:const_get)).grep(Class)
+        aggregate_list = classes.select { |item| item.include?(Sandthorn::AggregateRoot) }
+
+        modules = bounded_context_module.constants.map(&bounded_context_module.method(:const_get)).grep(Module).delete_if do |mo| 
+          classes.include?(mo) || mo == Sandthorn::BoundedContext::ClassMethods
+        end
+        
+        modules.each do |mo|
+          aggregate_list.concat(p_aggregate_list(mo))
+        end
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: We could use flatmap here: `aggregate_list += modules.flat_map (...)`

File: lib/sandthorn/bounded_context.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      def aggregate_list
+        @aggregate_list = p_aggregate_list(self)
+      end
+
+      private
+      
+      def p_aggregate_list(bounded_context_module)
+        return [] unless bounded_context_module.respond_to?(:constants)
+        
+        classes = bounded_context_module.constants.map(&bounded_context_module.method(:const_get)).grep(Class)
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: `classes` and `modules` could be methods?

File: lib/sandthorn/bounded_context.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        classes = get_classes(bounded_context_module)
+        aggregate_list = classes.select { |item| item.include?(Sandthorn::AggregateRoot) }
+        modules = get_modules(bounded_context_module, classes)
+        
+        aggregate_list += modules.flat_map { |m| p_aggregate_list(m) }
+        
+        aggregate_list
+      end
+
+      def get_classes namespace
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Suggestion rename to `classes_in(namespace)`, and similar for `get_modules`. Reason: `get_classes` implies a java-style getter (i.e. gets something on itself), which this isn't.

File: lib/sandthorn/bounded_context.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        namespace.constants.map(&namespace.method(:const_get)).grep(Class)
```
Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: By the way I really like this idiom, thanks for giving me the opportunity to learn it!

Comment (hallgren) [inline lib/sandthorn/bounded_context.rb]: We should all thank stack overflow ;)

File: lib/sandthorn/bounded_context.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+module Sandthorn
+  module BoundedContext
+    module ClassMethods
+      def aggregate_list
```
Comment (hallgren) [inline lib/sandthorn/bounded_context.rb]: @jesjos What do you think about changing the aggregate_list name to only aggregates?

Comment (jesjos) [inline lib/sandthorn/bounded_context.rb]: Perhaps `aggregate_types`? It's not a list of aggregate instances, right?

Comment (hallgren) [inline lib/sandthorn/bounded_context.rb]: Good call, its a list of classes that includes AggregateRoot.

**Pull request merged**